File: 01_Longest_Valid_Parentheses/Solution.py
def longest_valid_parentheses(s):
    stack = []
    stack.append(-1)
    max_lenght = 0
    for i in range(len(s)):
        if(s[i] == "("):
            stack.append(i)
        else:
            stack.pop()
            if not stack:
                stack.append(i)
            else:
                lenght = i - stack[-1]
                max_lenght = max(max_lenght, lenght)
    return max_lenght

s = ")()())"
result = longest_valid_parentheses(s)
print(result)

File: 01_Longest_Valid_Parentheses/Solution.java
import java.util.Stack;

public class Solution {

    public static int longestValidParentheses(String s) {

        Stack<Integer> stack = new Stack<>();

        stack.push(-1);

        int maxLength = 0;

        for (int i = 0; i < s.length(); i++) {

            if (s.charAt(i) == '(') {

                stack.push(i);

            } else {

                stack.pop();

                if (stack.isEmpty()) {

                    stack.push(i);

                } else {

                    int length = i - stack.peek();

                    maxLength = Math.max(maxLength, length);
                }
            }
        }

        return maxLength;
    }

    public static void main(String[] args) {
        String s = ")()())";
        int result = longestValidParentheses(s);
        System.out.println(result);
    }
}